File: tst/expert_fps.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_window("Expert FPS", -1, NULL);

    // test 1: set.expert returns ms
    int ms = pico_set_expert(1, 40);
    assert(ms == 25);

    // test 2: fps=0 returns 0
    int ms0 = pico_set_expert(1, 0);
    assert(ms0 == 0);

    // test 3: frame timing with compensation
    pico_set_expert(1, 40);
    int t0 = pico_get_now();
    Pico_Event e;
    for (int i=0; i<4; i++) {
        pico_input_event(&e, PICO_EVENT_ANY);
    }
    int elapsed = pico_get_now() - t0;
    assert(elapsed>=100 && elapsed<=110);

    // test 4: getter returns fps via pointer
    int fps;
    pico_set_expert(1, 40);
    assert(pico_get_expert(&fps) == 1);
    assert(fps == 40);

    // test 5: disable clears fps
    pico_set_expert(0, 0);
    assert(pico_get_expert(&fps) == 0);
    assert(fps == 0);

    // test 6: getter with NULL fps
    pico_set_expert(1, 60);
    assert(pico_get_expert(NULL) == 1);

    // test 7: fps=-1 returns ms=0
    int ms1 = pico_set_expert(1, -1);
    assert(ms1 == 0);

    // test 8: getter returns fps=-1
    pico_set_expert(1, -1);
    assert(pico_get_expert(&fps) == 1);
    assert(fps == -1);

    // test 9: fps=-1 returns immediately (dt ~0)
    pico_set_expert(1, -1);
    Pico_Event e2;
    int dt = pico_input_event(&e2, PICO_EVENT_ANY);
    assert(dt <= 5);
    assert(e2.type == PICO_EVENT_NONE);

    // test 10: event returns dt with fps timing
    pico_set_expert(1, 40);
    Pico_Event e3;
    int dt2 = pico_input_event(&e3, PICO_EVENT_ANY);
    assert(dt2>=25 && dt2<=30);

    pico_init(0);
    return 0;
}

File: lua/tst/expert_fps.lua
require 'pico.check'

pico.init(true)
pico.set.window { title="Expert FPS" }

-- test 1: returns ms
local ms = pico.set.expert(true, 40)
assert(ms == 25)

-- test 2: fps=0 returns -1 (block forever)
local ms0 = pico.set.expert(true, 0)
assert(ms0 == -1)

-- test 3: frame timing
pico.set.expert(true, 40)
local t0 = pico.get.now()
for i = 1, 4 do
    pico.input.event()
end
local elapsed = pico.get.now() - t0
assert(elapsed>=100 and elapsed<=110)

-- test 4: backward compat (no fps arg)
local ms1 = pico.set.expert(true)
assert(ms1 == -1)

-- test 5: fps=true (as fast as possible) returns ms=0
local ms2 = pico.set.expert(true, true)
assert(ms2 == 0)

-- test 6: fps=false returns ms=-1
local ms3 = pico.set.expert(true, false)
assert(ms3 == -1)

-- test 7: fps=true returns immediately, evt=nil
pico.set.expert(true, true)
local e1, dt1 = pico.input.event()
assert(e1 == nil)
assert(dt1 <= 5)

-- test 8: evt,dt from fps=40 timeout
pico.set.expert(true, 40)
local e2, dt2 = pico.input.event()
assert(e2 == nil)
assert(dt2>=25 and dt2<=30)

pico.init(false)
